### PR TITLE
Implement RuntimeHelpers.SizeOf

### LINKED ISSUE #97344: [API Proposal]: Get the managed size of a System.Type instance
### Background and motivation

As part of the same problem as https://github.com/dotnet/runtime/issues/97341, we have a scenario where we have a `System.Type`, we cannot rewrite the code in an AOT-safe manner to represent the type as a generic argument, and we want to be able to call `Unsafe.SizeOf<T>` for our `System.Type` instance.

This issue proposes a new method to get the size of the instance of a managed type from its `RuntimeTypeHandle`.

### API Proposal

```diff
namespace System.Runtime.CompilerServices;

public class RuntimeHelpers
{
+    public static int SizeOf(RuntimeTypeHandle type);
}
```

### API Usage

```csharp
var abi_element_size = RuntimeHelpers.SizeOf(AbiType.TypeHandle);
var byte_length = length * abi_element_size;
m._array = Marshal.AllocCoTaskMem(byte_length);
m._marshalers = new object[length];
var element = (byte*)m._array.ToPointer();
for (int i = 0; i < length; i++)
{
    m._marshalers[i] = Marshaler<T>.CreateMarshaler(array[i]);
    Marshaler<T>.CopyAbi(m._marshalers[i], (IntPtr)element);
    element += abi_element_size;
}
```

### Alternative Designs

We could place the API on `System.Runtime.CompilerServices.Unsafe` next to `SizeOf<T>()`.

### Risks

_No response_

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes https://github.com/dotnet/runtime/issues/97344

Implement the SizeOf API with the same behavior as the `sizeof` IL instruction.

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs [hunk 1]
```diff
@@ -437,6 +437,32 @@ public static void FixedAddressValueTypeTest()
 
             Assert.Equal(fixedPtr1, fixedPtr2);
         }
+
+        [InlineArray(4)]
+        private struct Byte4
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs:442]:
```suggestion
        [InlineArray(3)]
        private struct Byte3
```
3 byte would be more interesting - to ensure that the size is not getting aligned up.

File: src/coreclr/vm/reflectioninvocation.cpp [hunk 2]
```diff
@@ -2120,3 +2120,18 @@ FCIMPL2_IV(Object*, ReflectionEnum::InternalBoxEnum, ReflectClassBaseObject* tar
     return OBJECTREFToObject(ret);
 }
 FCIMPLEND
+
+extern "C" int32_t QCALLTYPE ReflectionInvocation_SizeOf(QCall::TypeHandle pType)
+{
+    QCALL_CONTRACT;
```
Comment (jkotas) [inline src/coreclr/vm/reflectioninvocation.cpp:2126]:
```suggestion
    QCALL_CONTRACT_NO_GC_TRANSITION;
```

File: src/coreclr/vm/reflectioninvocation.cpp [hunk 3]
```diff
@@ -2120,3 +2120,18 @@ FCIMPL2_IV(Object*, ReflectionEnum::InternalBoxEnum, ReflectClassBaseObject* tar
...
+{
+    QCALL_CONTRACT_NO_GC_TRANSITION;
+
+    int32_t size = 0;
+
+    BEGIN_QCALL;
+
+    size = pType.AsTypeHandle().GetSize();
+
+    END_QCALL;
```
Comment (jkotas) [inline src/coreclr/vm/reflectioninvocation.cpp:2134]:
```suggestion
    size = pType.AsTypeHandle().GetSize();
```

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 4]
```diff
@@ -416,6 +416,24 @@ public static IntPtr AllocateTypeAssociatedMemory(Type type, int size)
...
+        [SuppressGCTransition]
+        private static partial int SizeOf(QCallTypeHandle handle);
+
+        public static int SizeOf(RuntimeTypeHandle type)
+        {
+            RuntimeTypeHandle typeLocal = type;
+            if (typeLocal.IsNullHandle())
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);
+
+            int result = SizeOf(new QCallTypeHandle(ref typeLocal));
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:430]:
```suggestion
            if (type.IsNullHandle())
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);

            int result = SizeOf(new QCallTypeHandle(ref type));
```
I do not see why we need `typeLocal`

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.NativeAot.cs [hunk 5]
```diff
@@ -361,6 +361,22 @@ public static void PrepareDelegate(Delegate d)
 
             return RuntimeImports.RhNewObject(mt);
         }
+
+        public static unsafe int SizeOf(RuntimeTypeHandle type)
+        {
+            RuntimeTypeHandle typeLocal = type;
+            if (typeLocal.IsNull)
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.NativeAot.cs:368]:
```suggestion
            if (type.IsNull)
```

File: src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs [hunk 6]
```diff
@@ -211,5 +211,20 @@ internal static bool ObjectHasReferences(object obj)
...
+        public static int SizeOf(RuntimeTypeHandle type)
+        {
+            RuntimeTypeHandle typeLocal = type;
+            if (typeLocal.Value == IntPtr.Zero)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);
+
+            if (Type.GetTypeFromHandle(type)!.IsGenericParameter)
+                throw new ArgumentException(SR.Arg_TypeNotSupported);
+
+            return SizeOf(new QCallTypeHandle(ref typeLocal));
```
Comment (jkotas) [inline src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs:227]:
```suggestion
            if (type.Value == IntPtr.Zero)
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);

            if (Type.GetTypeFromHandle(type)!.IsGenericParameter)
                throw new ArgumentException(SR.Arg_TypeNotSupported);

            return SizeOf(new QCallTypeHandle(ref type));
```

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs (hunk 4 above)
Comment (jkoritzinsky) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:430]: I wasn't sure if we needed it for the QCallTypeHandle. I'll remove it if it's not required.

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs [hunk 7]
```diff
@@ -437,6 +437,32 @@ public static void FixedAddressValueTypeTest()
...
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(uint).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(long).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(ulong).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(float).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(double).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(Byte3).TypeHandle));
+            Assert.Equal(nint.Size, RuntimeHelpers.SizeOf(typeof(void*).TypeHandle));
+            Assert.Equal(nint.Size, RuntimeHelpers.SizeOf(typeof(delegate* <void>).TypeHandle));
+            Assert.Throws<ArgumentNullException>(() => RuntimeHelpers.SizeOf(default));
+            Assert.ThrowsAny<ArgumentException>(() => RuntimeHelpers.SizeOf(typeof(List<>).GetGenericArguments()[0].TypeHandle));
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs:464]: Add a test for System.Void?

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs [hunk 8]
```diff
@@ -437,6 +437,32 @@ public static void FixedAddressValueTypeTest()
...
+            Assert.Equal(1, RuntimeHelpers.SizeOf(typeof(byte).TypeHandle));
+            Assert.Equal(2, RuntimeHelpers.SizeOf(typeof(short).TypeHandle));
+            Assert.Equal(2, RuntimeHelpers.SizeOf(typeof(ushort).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(int).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(uint).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(long).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(ulong).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(float).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(double).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(Byte3).TypeHandle));
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs:460]:
```suggestion
            Assert.Equal(3, RuntimeHelpers.SizeOf(typeof(Byte3).TypeHandle));
```

File: src/coreclr/vm/reflectioninvocation.cpp [hunk 9]
```diff
@@ -2120,3 +2120,14 @@ FCIMPL2_IV(Object*, ReflectionEnum::InternalBoxEnum, ReflectClassBaseObject* tar
...
+
+extern "C" int32_t QCALLTYPE ReflectionInvocation_SizeOf(QCall::TypeHandle pType)
+{
+    QCALL_CONTRACT_NO_GC_TRANSITION;
+
+    int32_t size = 0;
+
+    size = pType.AsTypeHandle().GetSize();
+
+    return size;
```
Comment (jkotas) [inline src/coreclr/vm/reflectioninvocation.cpp:2132]:
```suggestion
    return pType.AsTypeHandle().GetSize();
```

**Review by jkotas — APPROVED** [review]

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 10]
```diff
@@ -416,6 +416,27 @@ public static IntPtr AllocateTypeAssociatedMemory(Type type, int size)
...
+        [SuppressGCTransition]
+        private static partial int SizeOf(QCallTypeHandle handle);
+
+        public static unsafe int SizeOf(RuntimeTypeHandle type)
+        {
+            if (type.IsNullHandle())
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);
+
+            TypeHandle handle = type.GetNativeTypeHandle();
+            if (!handle.IsTypeDesc && handle.AsMethodTable()->IsGenericTypeDefinition)
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:430]: Would it be better to do this check on the unmanaged side? It is tightly coupled with the unmanaged implementation of TypeHandle::GetSize and how it behaves in corner cases.

Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:430]: I think the condition to throw on here should rather be `ContainsGenericVariables` to handle partially instantiated types, like `MyStruct<int, !0>`.

Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:430]: (Add a test.)

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs [hunk 11]
```diff
@@ -437,6 +437,42 @@ public static void FixedAddressValueTypeTest()
...
+            Assert.Equal(2, RuntimeHelpers.SizeOf(typeof(short).TypeHandle));
+            Assert.Equal(2, RuntimeHelpers.SizeOf(typeof(ushort).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(int).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(uint).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(long).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(ulong).TypeHandle));
+            Assert.Equal(4, RuntimeHelpers.SizeOf(typeof(float).TypeHandle));
+            Assert.Equal(8, RuntimeHelpers.SizeOf(typeof(double).TypeHandle));
+            Assert.Equal(3, RuntimeHelpers.SizeOf(typeof(Byte3).TypeHandle));
+            Assert.Equal(nint.Size, RuntimeHelpers.SizeOf(typeof(void*).TypeHandle));
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs:461]: byref test may be useful too.

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs (hunk 10 above)
Comment (jkoritzinsky) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:430]: I can move it to the unmanaged side if you'd prefer. I don't have a preference.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.NativeAot.cs [hunk 12]
```diff
@@ -361,6 +361,27 @@ public static void PrepareDelegate(Delegate d)
...
+
+        public static unsafe int SizeOf(RuntimeTypeHandle type)
+        {
+            if (type.IsNull)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);
+
+            MethodTable* mt = type.ToMethodTable();
+
+            if (mt->ElementType == EETypeElementType.Void
+                || mt->ContainsGenericParameters)
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.NativeAot.cs:373]:
```suggestion
                || mt->IsGenericTypeDefinition)
```
I expect that this is going to fail the build. Native AOT does not materialize MethodTables for these abstract types, and so it does not have `ContainsGenericParameters` property.

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs [hunk 13]
```diff
@@ -416,6 +416,32 @@ public static IntPtr AllocateTypeAssociatedMemory(Type type, int size)
...
+        private static partial int SizeOf(QCallTypeHandle handle);
+
+        /// <summary>
+        /// Get the size of an object of the given type.
+        /// </summary>
+        /// <param name="type">The type to get the size of.</param>
+        /// <returns>The size of instances of the type.</returns>
+        /// <exception cref="ArgumentException">The passed-in type is not a valid type to get the size of.</exception>
+        /// <remarks>
+        /// This API has the same behavior as if you were to use the IL sizeof instruction with the passed in type as the operand.
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:431]: Or maybe even link to https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.sizeof that has more details.

File: src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs [hunk 14]
```diff
@@ -258,5 +258,29 @@ internal static bool ObjectHasReferences(object obj)
...
+        /// This API returns the same value as <see cref="Unsafe.SizeOf{T}"/> for the type that <paramref name="type"/> represents.
+        /// </remarks>
+        public static int SizeOf(RuntimeTypeHandle type)
+        {
+            if (type.Value == IntPtr.Zero)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.type);
+
+            Type typeObj = Type.GetTypeFromHandle(type)!;
+            if (typeObj.ContainsGenericParameters || typeObj.IsGenericParameter || typeObj == typeof(void))
+                throw new ArgumentException(SR.Arg_TypeNotSupported);
```
Comment (am11) [inline src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs:281]: Is this a limitation of mono runtime (which doesn't exist in coreclr)?

Comment (jkoritzinsky) [reply src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs:281]: No, this is the same limitation in CoreCLR and NativeAOT (though the implementations are slightly different due to how the different type systems are implemented).

**Pull request merged**